File: assignments of c/practise/1.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

#define MAX_VERTICES 1000
#define EPS 1e-10

typedef struct {
    double x, y;
} Point;

double polygonArea(Point *points, int n) {
    double area = 0.0;
    for (int i = 0; i < n; i++) {
        int j = (i + 1) % n;
        area += points[i].x * points[j].y;
        area -= points[j].x * points[i].y;
    }
    return fabs(area) / 2.0;
}

Point lineIntersection(Point p1, Point p2, Point p3, Point p4) {
    double x1 = p1.x, y1 = p1.y;
    double dx1 = p2.x - p1.x, dy1 = p2.y - p1.y;
    double x2 = p3.x, y2 = p3.y;
    double dx2 = p4.x - p3.x, dy2 = p4.y - p3.y;
    
    double denom = dx1 * dy2 - dy1 * dx2;
    
    Point result;
    if (fabs(denom) < EPS) {
        result = p2;
    } else {
        double t = ((x2 - x1) * dy2 - (y2 - y1) * dx2) / denom;
        result.x = x1 + t * dx1;
        result.y = y1 + t * dy1;
    }
    
    return result;
}

int isValidPolygon(Point *points, int n) {
    double area = polygonArea(points, n);
    if (area < EPS) return 0;
    
    for (int i = 0; i < n; i++) {
        int j = (i + 1) % n;
        double dx = points[j].x - points[i].x;
        double dy = points[j].y - points[i].y;
        double len = sqrt(dx * dx + dy * dy);
        if (len < 0.1 - EPS) return 0;
    }
    
    return 1;
}

int offsetPolygon(Point *points, int n, double offset, Point *result) {
    typedef struct {
        Point p1, p2;
    } Edge;
    
    Edge *edges = (Edge *)malloc(n * sizeof(Edge));
    
    for (int i = 0; i < n; i++) {
        int j = (i + 1) % n;
        double dx = points[j].x - points[i].x;
        double dy = points[j].y - points[i].y;
        double len = sqrt(dx * dx + dy * dy);
        
        if (len < EPS) {
            free(edges);
            return 0;
        }
        
        double nx = -dy / len;
        double ny = dx / len;
        
        edges[i].p1.x = points[i].x + nx * offset;
        edges[i].p1.y = points[i].y + ny * offset;
        edges[i].p2.x = points[j].x + nx * offset;
        edges[i].p2.y = points[j].y + ny * offset;
    }
    
    for (int i = 0; i < n; i++) {
        int j = (i + 1) % n;
        result[i] = lineIntersection(edges[i].p1, edges[i].p2, 
                                      edges[j].p1, edges[j].p2);
    }
    
    free(edges);
    return 1;
}

double calculateVolume(Point *points, int n, double h) {
    if (h <= 0) return 0.0;
    
    Point *offsetPoints = (Point *)malloc(n * sizeof(Point));
    
    if (!offsetPolygon(points, n, h, offsetPoints)) {
        free(offsetPoints);
        return 0.0;
    }
    
    if (!isValidPolygon(offsetPoints, n)) {
        free(offsetPoints);
        return 0.0;
    }
    
    double area = polygonArea(offsetPoints, n);
    free(offsetPoints);
    
    if (area <= 0) return 0.0;
    return area * h;
}

double minEdgeLength(Point *points, int n) {
    double minLen = 1e9;
    for (int i = 0; i < n; i++) {
        int j = (i + 1) % n;
        double dx = points[j].x - points[i].x;
        double dy = points[j].y - points[i].y;
        double len = sqrt(dx * dx + dy * dy);
        if (len < minLen) minLen = len;
    }
    return minLen;
}

double findMaxVolume(Point *points, int n) {
    double maxH = minEdgeLength(points, n) / 2.0;
    
    double maxVolume = 0.0;
    double bestH = 0.0;
    
    for (double h = 0.1; h <= maxH; h += 0.1) {
        double vol = calculateVolume(points, n, h);
        
        if (vol > maxVolume) {
            maxVolume = vol;
            bestH = h;
        }
    }
    
    return maxVolume;
}

int main() {
    int n;
    Point points[MAX_VERTICES];
    
    scanf("%d", &n);
    
    for (int i = 0; i < n; i++) {
        scanf("%lf %lf", &points[i].x, &points[i].y);
    }
    
    double maxVolume = findMaxVolume(points, n);
    
    printf("%.2f\n", maxVolume);
 
    
    
    return 0;
}

File: assignments of c/practise/1.py
import math

def calc_area(pts):
    tot = 0
    for i in range(len(pts)):
        j = (i + 1) % len(pts)
        tot += pts[i][0] * pts[j][1]
        tot -= pts[j][0] * pts[i][1]
    return abs(tot) / 2

def get_dir(pts):
    s = 0
    for i in range(len(pts)):
        j = (i + 1) % len(pts)
        s += pts[i][0] * pts[j][1] - pts[j][0] * pts[i][1]
    return -1 if s > 0 else 1

def shrink(pts, h, d):
    edges = []
    for i in range(len(pts)):
        p1 = pts[i]
        p2 = pts[(i + 1) % len(pts)]
        dx = p2[0] - p1[0]
        dy = p2[1] - p1[1]
        
        if dx == 0:
            nx = -1 if (dy > 0 and d == -1) or (dy < 0 and d == 1) else 1
            ny = 0
        else:
            nx = 0
            ny = 1 if (dx > 0 and d == -1) or (dx < 0 and d == 1) else -1
        
        edges.append(((p1[0] + nx * h, p1[1] + ny * h), (p2[0] + nx * h, p2[1] + ny * h)))
    
    new_pts = []
    for i in range(len(pts)):
        prev = (i - 1) % len(pts)
        a1, a2 = edges[prev]
        b1, b2 = edges[i]
        
        dx1 = a2[0] - a1[0]
        dy1 = a2[1] - a1[1]
        dx2 = b2[0] - b1[0]
        dy2 = b2[1] - b1[1]
        
        x1, y1 = a1[0], a1[1]
        x2, y2 = a2[0] + dx1 * 1000, a2[1] + dy1 * 1000
        x3, y3 = b1[0] - dx2 * 1000, b1[1] - dy2 * 1000
        x4, y4 = b2[0], b2[1]
        
        denom = (x1 - x2) * (y3 - y4) - (y1 - y2) * (x3 - x4)
        
        if abs(denom) < 0.0000000001:
            new_pts.append(a2)
        else:
            t = ((x1 - x3) * (y3 - y4) - (y1 - y3) * (x3 - x4)) / denom
            new_pts.append((x1 + t * (x2 - x1), y1 + t * (y2 - y1)))
    
    return new_pts

def check_ok(pts, h):
    for i in range(len(pts)):
        p1 = pts[i]
        p2 = pts[(i + 1) % len(pts)]
        length = abs(p2[0] - p1[0]) + abs(p2[1] - p1[1])
        if length - 2 * h < 0.099999:
            return False
    return True

n = int(input())
poly = []
for _ in range(n):
    x, y = map(int, input().split())
    poly.append((float(x), float(y)))

min_len = 999999
for i in range(n):
    p1 = poly[i]
    p2 = poly[(i + 1) % n]
    l = abs(p2[0] - p1[0]) + abs(p2[1] - p1[1])
    if l < min_len:
        min_len = l

max_h = (min_len - 0.1) / 2
if max_h < 0:
    max_h = 0
if max_h > 50:
    max_h = 50

steps = int(max_h * 10) + 1
if steps > 1000:
    steps = 1000

best = 0
direction = get_dir(poly)

for s in range(steps + 1):
    height = s * 0.1
    if height > max_h:
        break
    
    if check_ok(poly, height):
        new_poly = shrink(poly, height, direction)
        if len(new_poly) >= 3:
            a = calc_area(new_poly)
            if a > 0:
                v = a * height
                if v > best:
                    best = v

print(f"{best:.2f}")